### PR TITLE
[Merged by Bors] - Feat (GroupTheory/GroupAction/Hom/Pointwise) : generalize smul set lemmas to group actions

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
This is a generalization of `Mathlib/Algebra/Module/LinearMap/Pointwise.lean` from `LinearMapClass` to `MulActionHomClass`.

The preexisting lemmas are generalized.

- `image_smul_setₛₗ` : under a `σ`-equivariant map,
one has `h '' (c • s) = (σ c) • h '' s`.

- `preimage_smul_setₛₗ'` is a general version of the equality `h ⁻¹' (σ c • s) = c • h⁻¹' s`. It requires that `c` acts surjectively and `σ c` acts injectively.
It is provided with specific versions:
- `preimage_smul_setₛₗ_of_units` requires that `c` and `σ c` are units
- `MonoidHom.preimage_smul_setₛₗ` requires that `σ` is a `MonoidHom` and `c` is a unit
- `MonoidHomClass.preimage_smul_setₛₗ` requires that `σ` belongs to a `MonoidHomClass`and that `c` is a unit
- `Group.preimage_smul_setₛₗ` requires that the types of `c` and `σ c` are groups

- `image_smul_set`, `preimage_smul_set` and `Group.preimage_smul_set` are
the variants when `σ` is the identity.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+theorem MulAction.smul_bijective_of_is_unit
+    {M : Type*} [Monoid M] {α : Type*} [MulAction M α] {m : M} (hm : IsUnit m) :
+    Function.Bijective (fun (a : α) ↦ m • a) := by
+  lift m to Mˣ using hm
+  rw [Function.bijective_iff_has_inverse]
+  use (fun (a : α) ↦ m⁻¹ • a)
+  constructor
+  intro x; simp [← Units.smul_def]
+  intro x; simp [← Units.smul_def]
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/GroupAction/Pointwise.lean:41]:
```suggestion
  · intro x; simp [← Units.smul_def]
  · intro x; simp [← Units.smul_def]
```

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+-/
+
+open Set Pointwise
+
+theorem MulAction.smul_bijective_of_is_unit
+    {M : Type*} [Monoid M] {α : Type*} [MulAction M α] {m : M} (hm : IsUnit m) :
+    Function.Bijective (fun (a : α) ↦ m • a) := by
+  lift m to Mˣ using hm
+  rw [Function.bijective_iff_has_inverse]
+  use (fun (a : α) ↦ m⁻¹ • a)
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/GroupAction/Pointwise.lean:38]: I expect this works:
```suggestion
  use fun a : α ↦ m⁻¹ • a
```

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  simp only [← image_smul, image_image, map_smulₛₗ h]
+#align image_smul_setₛₗ image_smul_setₛₗ
+
+/-- Translation of preimage is contained in preimage of translation -/
+theorem smul_preimage_set_leₛₗ
+    [FunLike F M N] [MulActionSemiHomClass F σ M N]
+    (c : R) (t : Set N) :
+    c • h ⁻¹' t ⊆ h ⁻¹' (σ c • t) := by
+  rintro x ⟨y, hy, rfl⟩
+  refine ⟨h y, hy, (by rw [map_smulₛₗ])⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/GroupAction/Pointwise.lean:65]:
```suggestion
  exact ⟨h y, hy, by rw [map_smulₛₗ]⟩
```

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    h ⁻¹' (σ c • s) = c • h ⁻¹' s := by
+  apply le_antisymm
+  · intro m
+    obtain ⟨m', rfl⟩ := hc m
+    rintro ⟨n, hn, hn'⟩
+    refine ⟨m', ?_, rfl⟩
+    rw [map_smulₛₗ] at hn'
+    rw [mem_preimage, ← hc' hn']
+    exact hn
+  · exact smul_preimage_set_leₛₗ  M N σ h c s
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/GroupAction/Pointwise.lean:82]:
```suggestion
  · exact smul_preimage_set_leₛₗ M N σ h c s
```

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+/-!
+# Pointwise actions of equivariant maps
+
+- `image_smul_setₛₗ` : under a `σ`-equivariant map,
+  one has `h '' (c • s) = (σ c) • h '' s`.
+
+- `preimage_smul_setₛₗ'` is a general version of the equality
+  `h ⁻¹' (σ c • s) = c • h⁻¹' s`.
+  It requires that `c` acts surjectively and `σ c` acts injectively and
+is provided with specific versions:
```
Comment (jcommelin) [inline Mathlib/GroupTheory/GroupAction/Pointwise.lean:20]:
```suggestion
  is provided with specific versions:
```

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  - `preimage_smul_setₛₗ_of_units` when `c` and `σ c` are units
+  - `preimage_smul_setₛₗ` when `σ` belongs to a `MonoidHomClass`and `c` is a unit
+  - `MonoidHom.preimage_smul_setₛₗ` when `σ` is a `MonoidHom` and `c` is a unit
+  - `Group.preimage_smul_setₛₗ` : when the types of `c` and `σ c` are groups.
+
+- `image_smul_set`, `preimage_smul_set` and `Group.preimage_smul_set` are
+the variants when `σ` is the identity.
```
Comment (jcommelin) [inline Mathlib/GroupTheory/GroupAction/Pointwise.lean:27]:
```suggestion
  the variants when `σ` is the identity.
```

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+/-- `preimage_smul_setₛₗ` in the context of a groups -/
+theorem Group.preimage_smul_setₛₗ
+    {R S : Type*} [Group R] [Group S] (σ : R → S)
+    [MulAction R M] [MulAction S N]
+    {F : Type*} [FunLike F M N] [MulActionSemiHomClass F σ M N] (h : F)
+    (c : R) (s : Set N) :
+    h ⁻¹' (σ c • s) = c • h ⁻¹' s :=
+ preimage_smul_setₛₗ_of_units _ _ _ _ (Group.isUnit _) (Group.isUnit _) s
+
+variable (R)
```
Comment (jcommelin) [inline Mathlib/GroupTheory/GroupAction/Pointwise.lean:119]: Could the remainder of this file go in a section with the appropriate `variable`s?

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+variable [Monoid R] [Monoid S] (σ : R → S)
+variable [MulAction R M] [MulAction S N] [MulAction R M₁] [MulAction R M₂]
+variable {F : Type*} (h : F)
+
+-- @[simp] -- In #8386, the `simp_nf` linter complains:
+-- "Left-hand side does not simplify, when using the simp lemma on itself."
+-- For now we will have to manually add `image_smul_setₛₗ _` to the `simp` argument list.
+-- TODO: when lean4#3107 is fixed, mark this as `@[simp]`.
+theorem image_smul_setₛₗ [FunLike F M N] [MulActionSemiHomClass F σ M N]
```
Comment (jcommelin) [inline Mathlib/GroupTheory/GroupAction/Pointwise.lean:53]: Could you please put the next few lemmas in a section with
```lean
variable [FunLike F M N] [MulActionSemiHomClass F σ M N]
variable (c : R) (s : set M) (t : Set N)
```

**Comment by jcommelin** [comment]
Please do with my comments as you see fit. LGTM

bors d+

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean (hunk 7 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Pointwise.lean:119]: Yes!

File: Mathlib/GroupTheory/GroupAction/Pointwise.lean (hunk 8 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Pointwise.lean:53]: Yes!

**Pull request closed without merging**